sdk/core/azure-core/src/http/winhttp/win_http_transport.cpp

```diff
@@ -4,18 +4,572 @@
+      sessionHandle, connectionHandle, requestHandle, contentLength);
+
+  // Allocate the instance of the response on the heap with a shared ptr so this memory gets
+  // delegated outside the transport and will be eventually released.
+  auto rawResponse
+      = std::make_unique<RawResponse>(majorVersion, minorVersion, httpStatusCode, reasonPhrase);
+
+  rawResponse->SetBodyStream(std::move(stream));
+  rawResponse->AddHeaders(responseHeaders);
+
+  return rawResponse;
+}
+
+// Read the response from the sent request.
+int64_t Details::WinHttpStream::Read(Context const& context, uint8_t* buffer, int64_t count)
 {
-  void(context);
-  void(request);
+  context.ThrowIfCanceled();
+
+  if (count <= 0 || this->m_isEOF)
+  {
+    return 0;
+  }
+
+  int64_t totalNumberOfBytesRead = 0;
+  DWORD numberOfBytesRead = 0;
+
+  // Keep checking for data until there is nothing left.
+  do
+  {
+    context.ThrowIfCanceled();
+
+    // Check for available data.
+    DWORD numberOfBytesAvailable = 0;
+    if (!WinHttpQueryDataAvailable(this->m_requestHandle, &numberOfBytesAvailable))
+    {
+      // Errors include:
+      // ERROR_WINHTTP_CONNECTION_ERROR
+      // ERROR_WINHTTP_INCORRECT_HANDLE_STATE
+      // ERROR_WINHTTP_INCORRECT_HANDLE_TYPE
+      // ERROR_WINHTTP_INTERNAL_ERROR
+      // ERROR_WINHTTP_OPERATION_CANCELLED
+      // ERROR_WINHTTP_TIMEOUT
+      // ERROR_NOT_ENOUGH_MEMORY
+
+      DWORD error = GetLastError();
+      throw Azure::Core::Http::TransportException(
+          "Error while querying how much data is available to read. Error Code: "
+          + std::to_string(error) + ".");
+    }
+
+    context.ThrowIfCanceled();
+
+    DWORD numberOfBytesToRead = numberOfBytesAvailable;
+    if (numberOfBytesAvailable > count)
+    {
+      numberOfBytesToRead = static_cast<DWORD>(count);
+    }
+
+    if (!WinHttpReadData(
+            this->m_requestHandle,
+            (LPVOID)(buffer + totalNumberOfBytesRead),
+            numberOfBytesToRead,
+            &numberOfBytesRead))
+    {
+      // Errors include:
+      // ERROR_WINHTTP_CONNECTION_ERROR
+      // ERROR_WINHTTP_INCORRECT_HANDLE_STATE
+      // ERROR_WINHTTP_INCORRECT_HANDLE_TYPE
+      // ERROR_WINHTTP_INTERNAL_ERROR
+      // ERROR_WINHTTP_OPERATION_CANCELLED
+      // ERROR_WINHTTP_RESPONSE_DRAIN_OVERFLOW
+      // ERROR_WINHTTP_TIMEOUT
+      // ERROR_NOT_ENOUGH_MEMORY
+
+      DWORD error = GetLastError();
+      throw Azure::Core::Http::TransportException(
+          "Error while querying how much data is available to read. Error Code: "
+          + std::to_string(error) + ".");
+    }
+
+    totalNumberOfBytesRead += numberOfBytesRead;
+    count -= numberOfBytesRead;
+
+    if (numberOfBytesRead == 0
+        || (this->m_contentLength != -1
+            && this->m_streamTotalRead == this->m_contentLength - totalNumberOfBytesRead))
+    {
+      this->m_isEOF = true;
+      break;
+    }
+
+  } while (count > 0);
```

Comment:
We don't need to do this, just return whatever was read from the wire.
If we call bodyStream.Read(context, buffer, 100);
There is no contract about we must return 100 bytes, so we can return 10 and it is fine.
User fill keep calling Read until we return 0.

So, we don't need to keep reading from the wire until we have read `count`.
just return whatever was available (if that's less than count) or count (if that's less or equal to the available)

Reply:
Why are we pushing this requirement to the caller instead?

Reply:
At the contrary, we are doing this in favor to the caller.
We are returning the fastest we can so the caller can start using the data (streaming it).  
And this is the general contract from a bodyStream, the Read() function can return less than what you requested, that's why it returns the number of bytes it Read().

Returning whatever is available also gives some time to the winHttpClient to pull more data from the network (re-fill its own buffer), otherwise, you are spinning the CPU and demanding more bytes from the wire asap.